### DISCUSSION TITLE
Why does `ddlog -i tutorial.dl` generate 78 files / 27k sloc?

### DISCUSSION THREAD

**Discussion by infogulch** [vmware/differential-datalog#1046] [General]
Hi! I've never used ddlog before but it looks pretty neat based on the readme. I wanted to browse the rust code that is generated based on a .dl schema, but it seems that's not kept checked-in to this repo so I had to download and generate it myself. I guess I understand why...

```sh
➜ ddlog -i tutorial.dl    # note: the --no-staticlib option seemed to make no difference
```

```
➜ cloc tutorial_ddlog/
      81 text files.
      80 unique files.                              
       4 files ignored.

github.com/AlDanial/cloc v 1.82  T=0.07 s (1142.5 files/s, 485689.4 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Rust                            59           2380           2220          26305
TOML                            17             70             12            488
C/C++ Header                     1            139           1233            263
C                                1              5              6             37
-------------------------------------------------------------------------------
SUM:                            78           2594           3471          27093
-------------------------------------------------------------------------------
```

```
➜ find tutorial_ddlog/ -type d
tutorial_ddlog/
tutorial_ddlog/src
tutorial_ddlog/types
tutorial_ddlog/types/ddlog_std
tutorial_ddlog/types/inspect_log
tutorial_ddlog/types/internment
tutorial_ddlog/types/ddlog_log
tutorial_ddlog/types/vec
tutorial_ddlog/types/ddlog_rt
tutorial_ddlog/types/debug
tutorial_ddlog/types/ddlog_bigint
tutorial_ddlog/types/group
tutorial_ddlog/ovsdb
tutorial_ddlog/differential_datalog
tutorial_ddlog/differential_datalog/src
tutorial_ddlog/differential_datalog/src/render
tutorial_ddlog/differential_datalog/src/flatbuf
tutorial_ddlog/differential_datalog/src/dataflow
tutorial_ddlog/differential_datalog/src/record
tutorial_ddlog/differential_datalog/src/api
tutorial_ddlog/differential_datalog/src/program
tutorial_ddlog/differential_datalog/src/ddval
tutorial_ddlog/cmd_parser
tutorial_ddlog/.cargo
tutorial_ddlog/ddlog_derive
tutorial_ddlog/ddlog_derive/src
tutorial_ddlog/differential_datalog_test
```

It seems all the relevant code is in `./src/lib.rs` and `./types/tutorial.rs` which are a few thousand loc each, totally expected for code generated from such a comprehensive tutorial (over 750 lines by itself!).

What confuses me is that none of these other libraries seem to be directly related to the `tutorial.dl` file, but instead look like recursively vendored dependencies for the generated code. Why are dependencies vendored as part of the output?

**Comment by ryzhyk** [comment]
You're right.  All this code that is not auto-generated should live either on `crates.io` or in the DDlog installation directory.  This is on my todo list :)

**Reply by infogulch** [reply]
Ok, cool, thanks for responding. 😄

Have you considered trying to generate code on-the-fly during compilation by reading the .dl file in `build.rs`? E.g. [build script examples](https://doc.rust-lang.org/cargo/reference/build-script-examples.html). That would be Pretty Neat (tm) if the ddlog code is intended to be integrated as a library into a bigger project.

**Reply by ryzhyk** [reply]
We tend to write shell scripts or use `cargo make`, but `build.rs` might work too.
